### PR TITLE
Hash numpy arrays and pandas dataframes in tokenize

### PR DESCRIPTION
**Pull request by mrocklin**
This provides value-based hashing for numpy arrays and pandas dataframes.  This should help with consistent hashing when creating tokens in dask graphs.  This converts the objects into buffers and then uses `md5`.

cc @jcrist 

```python
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: A = pd.DataFrame({'x': np.arange(1000000), 'y': list('abcdefghij'*100000)})
In [4]: B = pd.DataFrame({'x': np.arange(1000000), 'y': list('abcdefghij'*100000)})

In [5]: from dask.base import tokenize

In [6]: %time tokenize(A)
CPU times: user 45.8 ms, sys: 593 µs, total: 46.4 ms
Wall time: 45.7 ms
Out[6]: '53bd8832e0cc1c8a2da346ba0b7d4685'

In [7]: %time tokenize(B)
CPU times: user 40.3 ms, sys: 298 µs, total: 40.6 ms
Wall time: 39.9 ms
Out[7]: '53bd8832e0cc1c8a2da346ba0b7d4685'
```

### CODE REVIEW

File: dask/base.py [hunk 1]
```diff
@@ -132,14 +132,42 @@ def normalize_function(func):
...
         lambda a: normalize_function(a) if callable(a) else a)
 normalize_token.register(dict, lambda a: tuple(sorted(a.items())))
+
 with ignoring(ImportError):
     import pandas as pd
-    normalize_token.register(pd.DataFrame,
-            lambda a: (id(a), len(a), list(a.columns)))
-    normalize_token.register(pd.Series, lambda a: (id(a), len(a), a.name))
+
+    @partial(normalize_token.register, pd.Index)
```
Comment (jcrist) [inline dask/base.py]: This is a clever use of `partial`.

File: dask/base.py [hunk 2]
```diff
@@ -132,14 +132,42 @@ def normalize_function(func):
...
+    def normalize_index(ind):
+        return tokenize(ind.name, ind.values)
+
+    @partial(normalize_token.register, pd.Categorical)
+    def normalize_categorical(cat):
+        return tokenize(cat.codes, cat.categories, cat.ordered)
+
+    @partial(normalize_token.register, pd.Series)
+    def normalize_series(s):
+        return tokenize(s.name, s._data.blocks[0].values, s.dtype, s.index)
```
Comment (jcrist) [inline dask/base.py]: Are you sure you want to call `tokenize` internally in these? That's going to apply `normalize_token` to each of the arguments, then string the tuple, then md5 it. If the arguments are already normalized, you could probably just return the tuple instead.

Comment (mrocklin) [inline dask/base.py]: Often I want the recursion.  For example in the DataFrame function we use numpy arrays.  Calling tokenize might be slightly less efficient but it's less error-prone for me

Comment (jcrist) [inline dask/base.py]: Perhaps just `map(normalize_token, ...)` (or better yet, only on the ones known to need normalizing) instead to avoid the str + hash? Not mandatory, just a thought.

Comment (mrocklin) [inline dask/base.py]: str and hash are mostly free

```python
In [1]: from hashlib import md5

In [2]: %time _ = md5(str(['1y9hfshfs90f8hfs0shf', 'hf90s8hfsh90h098hf3']))
CPU times: user 9 µs, sys: 2 µs, total: 11 µs
Wall time: 11.9 µs

In [3]: %timeit md5(str(['1y9hfshfs90f8hfs0shf', 'hf90s8hfsh90h098hf3']))
The slowest run took 20.88 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 1.05 µs per loop
```

Comment (mrocklin) [inline dask/base.py]: Ah, although I suppose that this doesn't count extra calls to all of the tokenize machinery.

```python
In [1]: from dask.base import tokenize

In [2]: %time _ = tokenize(['1y9hfshfs90f8hfs0shf', 'hf90s8hfsh90h098hf3'])
CPU times: user 17 µs, sys: 2 µs, total: 19 µs
Wall time: 22.2 µs

In [3]: %timeit tokenize(['1y9hfshfs90f8hfs0shf', 'hf90s8hfsh90h098hf3'])
The slowest run took 7.86 times longer than the fastest. This could mean that an intermediate result is being cached 
100000 loops, best of 3: 2.79 µs per loop
```

**Comment by mrocklin** [comment]
This finally passes tests.  There were a number of odd corner cases.

File: dask/base.py (hunk 2 above)
Comment (mrocklin) [inline dask/base.py]: Now using `normalize_token` rather than `tokenize`

**Comment by mrocklin** [comment]
@jcrist anything else to do here?

**Comment by jcrist** [comment]
Looks good to me.

File: dask/base.py [hunk 3]
```diff
@@ -132,14 +132,50 @@ def normalize_function(func):
...
+        if x.dtype.hasobject:
+            try:
+                data = md5('-'.join(x.flat)).hexdigest()
+            except TypeError:
+                data = md5(b'-'.join([str(item).encode() for item in x.flat])).hexdigest()
+        else:
+            try:
+                data = md5(x.view('i1').data).hexdigest()
+            except (BufferError, AttributeError, ValueError):
+                data = md5(x.copy().view('i1').data).hexdigest()
```
Comment (jakirkham) [inline dask/base.py]: Sorry to necro an old thread. Just appears that the `i1` view has its origin here. Am curious why `u1` was not used for the view considering the intent?

Comment (mrocklin) [reply dask/base.py]: I suspect that we were trying to get a memoryview with a one-byte stride.  I don't currently recall exactly what broke without this.

Comment (jcrist) [reply dask/base.py]: Since we're just trying to view the data as bytes it really doesn't matter whether it's signed or unsigned, just that the element is just a single byte in size. `u1` would have been an equally valid choice, but there's nothing wrong with `i1` - they give the same results:

```python
In [22]: import numpy as np

In [23]: from hashlib import md5

In [24]: a = np.array([-1, 1, 2])

In [25]: md5(a.view('i1').data).hexdigest()
Out[25]: '2c2020a2d6693493847c4d25749156e2'

In [26]: md5(a.view('u1').data).hexdigest()
Out[26]: '2c2020a2d6693493847c4d25749156e2'
```

Comment (mrocklin) [reply dask/base.py]: Yeah, that would be fine.  Was `u1` causing an issue?

Comment (jakirkham) [reply dask/base.py]: Guessing you meant `i1`, not that I know of. Just always found it a bit easier to follow unsigned bytes instead of signed bytes.
